Introduce enum classes to identify and categorize events

rniwa opened the pull request:
#### 490b4fdda464141c8eedb0ec367dbf76c6077a77
<pre>
Introduce enum classes to identify and categorize events
<a href="https://bugs.webkit.org/show_bug.cgi?id=267188">https://bugs.webkit.org/show_bug.cgi?id=267188</a>

Reviewed by Yusuke Suzuki.

This PR introduces enum classes to uniquely identify and category builtin events.

EventType enum class has an unique value for each builtin event, and this allows various
functions to use a switch statement based on event types instead of having a sequence of
if-else statements.

EventCategory enum class bitfields identifies categories a given event belongs such as
TouchRelated and MouseMoveRelated. These categories are specified for each event in
EventNames.json by specifying &quot;categories&quot; array, which has been renamed from &quot;types&quot;.

This PR replaces EventNames&apos;s is*EventType functions with checks for the corresponding
categories so that we would do a single HashMap lookup and a bitfield check instead of
a bunch of comparisons against AtomString&apos;s.

EventTypeInfo encompasses these two enum classes for a given event type as well as a flag
indicating whether a given event has default event handlers (C++ code which responds to
a given event dispatch), which is also specified in EventNames.json. In the future this
object can be extended to work like QualifiedName but for events.

Because event type names are AtomString, which is a thread specific object, a map of
event types to EventTypeInfo is stored in each thread&apos;s EventNames object using
MemoryCompactLookupOnlyRobinHoodHashMap.

In addition, this PR introduces the counters for each event type and plugin elements at
Document level so that EventDispatcher can circumvent EventPath construction and/or calls
to default event handlers.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addListenerTypeIfNeeded):
(WebCore::Document::didAddEventListenersOfType): Added.
(WebCore::Document::didRemoveEventListenersOfType): Added.

* Source/WebCore/dom/Document.h:
(WebCore::Document::hasEventListnersOfType const): Added.
(WebCore::Document::hasConnectedPluginElements): Added.
(WebCore::Document::didConnectPluginElement): Added.
(WebCore::Document::didDisconnectPluginElement): Added.

* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::hasRelevantEventListener): Added.
(WebCore::resetAfterDispatchInShadowTree): Extracted from EventDispatcher::dispatchEvent.
(WebCore::EventDispatcher::dispatchEvent):

* Source/WebCore/dom/EventListenerMap.h:
(WebCore::EventListenerMap::enumerateEventListenerTypes): Added.

* Source/WebCore/dom/EventNames.json: Updated per the rename of &quot;types&quot; to &quot;categories&quot;,
and the introduction of &quot;defaultEventHandlers&quot; flag, which indicates whether a given event
has default event handler behaviors or not.

* Source/WebCore/dom/EventTarget.cpp:
(WebCore::EventTarget::addEventListener):
(WebCore::EventTarget::removeEventListener):
(WebCore::EventTarget::legacyTypeForEvent): Renamed from legacyType and declared in the
header file so that EventDispatcher can use it.
(WebCore::EventTarget::fireEventListeners):

* Source/WebCore/dom/EventTarget.h:

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::moveNodeToNewDocument):
(WebCore::isTouchRelatedEventType): Updated to use EventCategory bitfields.
(WebCore::tryAddEventListener): Notifies the document whenever an event listener is added.
(WebCore::tryRemoveEventListener): Ditto for removal.
(WebCore::Node::defaultEventHandler): Updated to use a switch statement and EventCategory
bitfields checks.
(WebCore::Node::willRespondToMouseMoveEvents const): Updated to use EventCategory checks.
(WebCore::Node::willRespondToTouchEvents const): Ditto.
(WebCore::Node::willRespondToMouseClickEventsWithEditability const): Ditto.

* Source/WebCore/dom/Node.h:

* Source/WebCore/dom/make-event-names.py:
(main): Now generates the code for EventType, EventCategory, and EventTypeInfo based on
the information newly available in EventNames.json.

* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::insertedIntoAncestor): Added to notify the document when
a plugin element gets connected.
(WebCore::HTMLPlugInElement::removedFromAncestor): Ditto for disconnect.

* Source/WebCore/html/HTMLPlugInElement.h:

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::didSecureTransitionTo):
(WebCore::LocalDOMWindow::addEventListener): Updated to use a switch statement and
EventCategory bitfields checks. Also notify the document of a new event listener.
(WebCore::LocalDOMWindow::removeEventListener): Ditto for the removal.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldMakeEventListenerPassive): Now takes EventTypeInfo to do
EventCategory bitfields checks.

* Source/WebCore/page/Quirks.h:

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::handleEvent):

Canonical link: <a href="https://commits.webkit.org/272801@main">https://commits.webkit.org/272801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ec1710b1f80d7ee263306094a1e1d7627f11e2d3

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32916 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11674 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34793 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35540 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29737 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/33867 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14014 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8850 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29180 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33372 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9837 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29395 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8572 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8719 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29351 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/36872 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29894 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29760 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34852 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8849 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6802 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32706 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10541 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9454 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4260 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9475 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/EventDispatcher.cpp:
```diff
@@ -172,6 +200,12 @@ void EventDispatcher::dispatchEvent(Node& node, Event& event)
         }
     }
 
+    if (hasNoEventListnerOrDefaultEventHandler) {
```
This code path is taken when the target or the related target is inside a shadow tree such that we must compute EventPath to figure out whether the final target / related target will be still inside a shadow tree or not.

Constellation commented on Source/WebCore/dom/Document.h:
```diff
@@ -2343,6 +2352,9 @@ class Document
     unsigned m_dataListElementCount { 0 };
 
     OptionSet<ListenerType> m_listenerTypes;
+    MemoryCompactRobinHoodHashMap<AtomString, unsigned> m_eventListenerCounts;
```
How many enum do we have? How about having EnumeratedArray instead?

rniwa replied:
As far as I can tell, using EnumeratedArray instead of MemoryCompactRobinHoodHashMap doesn't yield any additional benefit.

Constellation commented on Source/WebCore/dom/make-event-names.py:
```diff
@@ -209,7 +251,27 @@ def writeln(text):
             if conditional:
                 writeln('#endif')
-
+        writeln(f'    {delimiter} m_typeInfoMap({{')
```
In a subsequent change, we should make it tri-searching code instead (as the same to HTMLName etc.).

rniwa replied:
You think that would be beneficial even though the string is already literal / AtomString?

Constellation reviewed (approved):
r=me

nt1m commented on Source/WebCore/dom/EventNames.json:
```diff
@@ -1,5 +1,5 @@
 {
-    "DOMActivate": {"types": ["MouseClickRelated"]},
+    "DOMActivate": { "categories": ["MouseClickRelated"], "defaultEventHandler": true },
```
```suggestion
    "DOMActivate": {
        "categories": ["MouseClickRelated"],
        "defaultEventHandler": true
     },
```
nit: Can we format like this?

rniwa replied:
We can but I don't want each event to take up more than one line.

rniwa committed:
Introduce enum classes to identify and categorize events
https://bugs.webkit.org/show_bug.cgi?id=267188

Reviewed by Yusuke Suzuki.

This PR introduces enum classes to uniquely identify and category builtin events.

EventType enum class has an unique value for each builtin event, and this allows various
functions to use a switch statement based on event types instead of having a sequence of
if-else statements.

EventCategory enum class bitfields identifies categories a given event belongs such as
TouchRelated and MouseMoveRelated. These categories are specified for each event in
EventNames.json by specifying "categories" array, which has been renamed from "types".

This PR replaces EventNames's is*EventType functions with checks for the corresponding
categories so that we would do a single HashMap lookup and a bitfield check instead of
a bunch of comparisons against AtomString's.

EventTypeInfo encompasses these two enum classes for a given event type as well as a flag
indicating whether a given event has default event handlers (C++ code which responds to
a given event dispatch), which is also specified in EventNames.json. In the future this
object can be extended to work like QualifiedName but for events.

Because event type names are AtomString, which is a thread specific object, a map of
event types to EventTypeInfo is stored in each thread's EventNames object using
MemoryCompactLookupOnlyRobinHoodHashMap.

In addition, this PR introduces the counters for each event type and plugin elements at
Document level so that EventDispatcher can circumvent EventPath construction and/or calls
to default event handlers.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::addListenerTypeIfNeeded):
(WebCore::Document::didAddEventListenersOfType): Added.
(WebCore::Document::didRemoveEventListenersOfType): Added.

* Source/WebCore/dom/Document.h:
(WebCore::Document::hasEventListnersOfType const): Added.
(WebCore::Document::hasConnectedPluginElements): Added.
(WebCore::Document::didConnectPluginElement): Added.
(WebCore::Document::didDisconnectPluginElement): Added.

* Source/WebCore/dom/EventDispatcher.cpp:
(WebCore::hasRelevantEventListener): Added.
(WebCore::resetAfterDispatchInShadowTree): Extracted from EventDispatcher::dispatchEvent.
(WebCore::EventDispatcher::dispatchEvent):

* Source/WebCore/dom/EventListenerMap.h:
(WebCore::EventListenerMap::enumerateEventListenerTypes): Added.

* Source/WebCore/dom/EventNames.json: Updated per the rename of "types" to "categories",
and the introduction of "defaultEventHandlers" flag, which indicates whether a given event
has default event handler behaviors or not.

* Source/WebCore/dom/EventTarget.cpp:
(WebCore::EventTarget::addEventListener):
(WebCore::EventTarget::removeEventListener):
(WebCore::EventTarget::legacyTypeForEvent): Renamed from legacyType and declared in the
header file so that EventDispatcher can use it.
(WebCore::EventTarget::fireEventListeners):

* Source/WebCore/dom/EventTarget.h:

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::moveNodeToNewDocument):
(WebCore::isTouchRelatedEventType): Updated to use EventCategory bitfields.
(WebCore::tryAddEventListener): Notifies the document whenever an event listener is added.
(WebCore::tryRemoveEventListener): Ditto for removal.
(WebCore::Node::defaultEventHandler): Updated to use a switch statement and EventCategory
bitfields checks.
(WebCore::Node::willRespondToMouseMoveEvents const): Updated to use EventCategory checks.
(WebCore::Node::willRespondToTouchEvents const): Ditto.
(WebCore::Node::willRespondToMouseClickEventsWithEditability const): Ditto.

* Source/WebCore/dom/Node.h:

* Source/WebCore/dom/make-event-names.py:
(main): Now generates the code for EventType, EventCategory, and EventTypeInfo based on
the information newly available in EventNames.json.

* Source/WebCore/html/HTMLPlugInElement.cpp:
(WebCore::HTMLPlugInElement::insertedIntoAncestor): Added to notify the document when
a plugin element gets connected.
(WebCore::HTMLPlugInElement::removedFromAncestor): Ditto for disconnect.

* Source/WebCore/html/HTMLPlugInElement.h:

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::didSecureTransitionTo):
(WebCore::LocalDOMWindow::addEventListener): Updated to use a switch statement and
EventCategory bitfields checks. Also notify the document of a new event listener.
(WebCore::LocalDOMWindow::removeEventListener): Ditto for the removal.

* Source/WebCore/page/Quirks.cpp:
(WebCore::Quirks::shouldMakeEventListenerPassive): Now takes EventTypeInfo to do
EventCategory bitfields checks.

* Source/WebCore/page/Quirks.h:

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::handleEvent):

Canonical link: https://commits.webkit.org/272801@main
